vernac/vernacentries.ml

```diff
@@ -1351,7 +1351,8 @@ let expand filename =
 let vernac_add_loadpath ~implicit pdir coq_path =
   let open Loadpath in
   let pdir = expand pdir in
-  add_vo_path { unix_path = pdir; coq_path; has_ml = true; implicit; recursive = true }
+  add_vo_path { unix_path = pdir; coq_path; has_ml = true; implicit; recursive = true };
+  DebugHook.refresh_bpts ()
```

Comment:
This does not look right to unconditionally refresh breakpoints here. The debugger should be responsible for it, not some random code in Vernacentries. This would also be broken if some external plugin calls `add_vo_path` directly without going through a vernac command.

Reply:
I can certainly move the call into `add_vo_path`.  However, the call is needed to let users set breakpoints in files/modules loaded with `Add LoadPath` commands before they've been loaded.  CoqIDE passes absolute file names (that's all it knows about), which can't be translated into a module name until the loadpath is defined.  How else would you handle this?

Reply:
> I can certainly move the call into add_vo_path

This is even worse.

> It has to know when the loadpath is added. How else would you handle this?

Can't you simply keep a cache in the debugger (from Coq side) that remembers the set of loadpaths and performs the refresh when the user interacts with the debugger, e.g. adding breakpoints / launching a debug session?

Reply:
User interaction with the debugger in CoqIDE and execution of script commands are in different processes and asynchronous to each other.  Some call is necessary to get consistency.

Consider the use case in which the user runs CoqIDE and immediately sets a breakpoint in a module that's loaded with `Add LoadPath`, then runs the toplevel script.  The modpath is unknown until the `Add LoadPath` is executed later, which is asynchronous to any user interaction.  Absent the call, how would the debugger know when the path has been added?

Reply:
The breakpoint-checking code in the debugger could handle the updating, but that means checking for loadpath updates on every single step during execution.  The performance cost may be acceptable if there's a boolean that's set whenever the set of loadpaths is updated.  The debugger would have to be able to clear the flag, of course.

Reply:
There is something I really don't understand about what you're talking about. If you have a global non-synchronized state that is living its life in a distinct process, this is going to literally violate all the invariants of the coq process. In this view I don't see how it's even allowed to evaluate vernac commands on the fly while a debugger session is running. Anything you'd get after could be garbage.

Reply:
I was merely emphasizing that Coq (interpreting the script) and CoqIDE (responding to user actions) are separate processes.  Of course, Coq can't be in `Tactic_debug` or `Tacinterp` while it's processing a vernac command.  But Coq may enter and exit those modules frequently as the user steps through scripts with Forward/Ctrl-Down.  The breakpoints persist until explicitly removed by a call from CoqIDE.

Another approach is to call `refresh_bpts` in `Idetop.add` after processing each sentence.  That's a little less efficient than where the call is now because updates are only needed after `Add LoadPath`.

Let me know if it's still a bit unclear.

Reply:
Also note that users can add/remove breakpoints both when Coq is idle and when Coq is stopped at a breakpoint.

Reply:
> Let me know if it's still a bit unclear.

I still don't understand. The breakpoints are living in the coq process, not the coqide process. You only need them when interacting with the debugger, from the coqide process. Why can't you synchronize whenever any call to the debugger is performed, coq-side?

Reply:
Forgot to mention that the use of `Add LoadPath` in scripts is highly discouraged, so that I wouldn't even bother supporting this use case anyways.

Reply:
> Why can't you synchronize whenever any call to the debugger is performed, coq-side?

I mentioned that possibility [above](https://github.com/coq/coq/pull/14354#discussion_r691532421), though instead of adding the boolean flag, I could do this, which is slightly simpler:

```
IN loadpath.ml:
let get_load_paths () = !load_paths

IN tactic_debug.ml:
let prev_load_paths = ref []
let debug_prompt ... =
       :
    if LoadPath.get_load_paths () != !prev_load_paths then begin
      prev_load_paths := LoadPath.get_load_paths ();
      DebugHook.refresh_bpts ()
    end;
```

IIRC, `Set Ltac Debug` already has a performance cost of about 20% because it calls `debug_prompt` for every tactic in Ltac tactics.  I don't want it to get much higher, hence the need for the performance tweak.

> the use of `Add LoadPath` in scripts is highly discouraged, so that I wouldn't even bother supporting this use case anyways.

It significantly breaks the debugger for anyone using the command, which could lead to a bug report and IMHO would have to be documented.  It's not difficult to address this and code to do so is already in hand.  Which is worse--a user-visible bug/limitation or exposing a little more information in a lower-level module?

FWIW, the doc currently says (recently added in the coqc description) "The drawback of [Add LoadPath] is that it adds environment-specific information (the PATH) to your script, making it non-portable, so we discourage using this approach."

What do you think?